### DISCUSSION TITLE
understanding-scopes-with-ihttpclientfactory-message-handlers/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#52] [Announcements]
# understanding-scopes-with-ihttpclientfactory-message-handlers/

In this post I discuss how handler DI scopes work in IHttpClientFactory, and how they're different to the request DI scope.

https://andrewlock.net/understanding-scopes-with-ihttpclientfactory-message-handlers/

**Comment by nisar-ahmad** [comment]
Thanks for writing this great article. In my case, I can't use `HttpContext` since my `DelegatingHandler` (which adds bearer tokens) is run in the context of a hosted service when an event is received from service bus. I'm using `MassTransit` and my event handler runs inside a consumer wrapper that has `ConsumeContext`. Any idea how I can inject the `IServiceProvider` scoped to my custom 'event handling' scope? Basically, I send the Bearer Token received from the incoming event with the outgoing Http requests.

**Reply by nisar-ahmad** [reply]
I was able to solve this using `AsyncLocal` approach explained in one of the answers here:
https://stackoverflow.com/questions/69801603/obtaining-a-scoped-service-in-a-custom-delegating-handler-in-blazor-server

**Comment by pablocom** [comment]
Interesting, thanks for sharing. 
Note there are 3 typos where `IHttpClintFactory` is written

[a comment by an automated account is not shown]

**Comment by maksim83** [comment]
Great article!  Ther weak side is    var httpConext = _accessor.HttpContext; 
Under high load i started getting httpcontexts of random users. 

Morten Møller Riis  assumed such a possibility, i faced with it in practice.

**Comment by ghost** [comment]
Your post is great.
Missing superlatives here.
Clarity, progression logic, use cases, and a solution! All in one.

I understood by myself that scope shenanigans were going on when I had to "share state": it was not shared, and that was definitely surprising. I already knew about HttpFactory details from your other articles, but same as you: I did not see the full consequences... until I had to share state.

I would not have found the solution: resolving via service locator.

Thank you.

**Comment by GitSnafu** [comment]
Great article. Unfortunately it didn't address my specific problem, but it taught me enough to confirm some of my suspicions with regard to DI scopes and HttpClient handlers. In my case I want to inject a dependency into a custom handler where that dependency is configurable per HttpClient, so in a sense I need the dependency to be scoped by HttpClient, but the solution isn't to use DI scopes (or perhaps there is such a solution, but I found another arguably simpler one).

Basically, I have an extension method that allows you do the following (this from a unit test, hence the stubbed primary handler):

```CSharp
services.AddHttpClient("http-client-1")                                    
    .ConfigurePrimaryHttpMessageHandler(() => new HttpMessageHandlerStub())
    .AddRequestMetrics(opt => opt.WithFilter(httpRequest => httpRequest.RequestUri!.PathAndQuery.EndsWith("/path1")));
```

And an extension method that looked something like this (with various permutations attempted without success):

```CSharp
public static IHttpClientBuilder AddRequestMetrics(this IHttpClientBuilder builder, Action<IHttpClientRequestMetricsFilterOptions> options = null) 
{                                                                                                                                                  
    var httpClientRequestMetricsFilter = new HttpClientRequestMetricsFilter();                                                                     
                                                                                                                                                   
    options?.Invoke(httpClientRequestMetricsFilter);                                                                                               
                                                                                                                                                   
    // HTTP request middleware should be registered as transient                                                                                     
    builder.Services.TryAddTransient(sp => new HttpClientRequestMetricsHandler(sp.GetRequiredService<IMetrics>(), httpClientRequestMetricsFilter));
                                                                                                                                                   
    return builder.AddHttpMessageHandler<HttpClientRequestMetricsHandler>();                                                                       
}                                                                                                                                                  
```

However, all instances of the handler received the same instance of the HttpClientRequestMetricsFilter, either the first or the last one that was registered (I forget which now). I had somehow hoped that registering something with IHttpClientBuilder.Services would scope it to the corresponding HttpClient being configured, but that's not the case.

Instead the solution I came up with was to do this:

```CSharp
public static IHttpClientBuilder AddRequestMetrics(this IHttpClientBuilder builder, Action<IHttpClientRequestMetricsFilterOptions> options = null) 
{                                                                                                                                                  
    var httpClientRequestMetricsFilter = new HttpClientRequestMetricsFilter();

    options?.Invoke(httpClientRequestMetricsFilter);

    return builder.AddHttpMessageHandler(sp => new HttpClientRequestMetricsHandler(sp.GetRequiredService<IMetrics>(), httpClientRequestMetricsFilter));
}                                                                                                                                                  
```

Since the handler is indeed scoped to the HttpClient being configured each HttpClient gets a handler with the speficic HttpClientRequestMetricsFilter configured for that client.

Hopefully this will help someone.

**Comment by SergeyDevyatko** [comment]
Brilliant! I'm a big fun of your articles)

**Comment by GSDan** [comment]
Thank you, this solved a real headache for me.

**Comment by storey247** [comment]
Thanks again for this post, many years later, as it helped me identify an issue I was having with LOH and GC Handles growing exponentially. 

One other catch 22 of this issue that may be worth adding to your “does it matter” section is if the Scoped dependency relies on `IDisposable` pattern 🤦‍♂️

My scoped injection into my handler did this because it made use of a semaphore… scoped injections getting trapped by these handlers meant that disposal wasn’t happening when I expected it to and boom. 

Just sharing above in case this catches anyone else out and to thank you again for this great post 🙏
